### PR TITLE
fix: extended query anonymizer tests with functional tests queries

### PR DESCRIPTION
**Pull request by tolgadur**
### Description 
This tests take queries used in the ksqldb functional tests to test the [QueryAnonymizer](https://confluentinc.atlassian.net/browse/KCI-341) and make it more robust before shipping. The issues identified and fixed are:

- Types weren't anonymized correctly (i.e. custom types weren't anonymized at all)
- PRIMARY KEYS / KEYS were omitted in the anonymized query 
- Explain queries weren't anonymized correctly. I.e. `EXPLAIN SELECT * FROM my_table;` just resulted in `EXPLAIN query;`
- If the same name was used to create a table and stream name they used to result in the same name. (not sure if this is expected behavior?)
- SELECT statements werent anonymized at all.
- \* was completely omitted in the anonymizations. i.e. `SELECT * FROM my_table`
- JOIN statements were completely ommited in anonymizations. 

Unfortunately with the current implementation I did not find a way to determine whether it's a stream or table in SELECT statements. Thus, the anonymizer is just using the generic 'source'. I.e. `SELECT * FROM my_table` becomes `SELECT * FROM source1`.

I was also not able to correctly determine the keys in JOIN statements. Thus, the anonymizer is just using the generic `key1=key2`. I.e. 
```
INSERT INTO OUTPUT SELECT col1, col2, col3  FROM SOURCE1 S1 JOIN SOURCE2 S2 WITHIN 1 SECOND ON col1.k=col2.k;
``` 
becomes 
```
INSERT INTO OUTPUT SELECT col1, col2, col3 FROM SOURCE1 S1 JOIN SOURCE2 S2 WITHIN 1 SECOND ON col1.k=col2.k;
``` 

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-parser/src/test/java/io/confluent/ksql/util/QueryAnonymizerTest.java [hunk 1]
```diff
@@ -43,7 +43,7 @@ public void shouldWorkAsExpectedWhenPassedAParseTreeInsteadOfString() {
   }
 
   @Test
-  public void shouldLeaveListAndPrintStatementsAsTheyAre() {
+  public void shouldLeaveListAndPrintStatementsAsTheyAre() throws Exception {
```
Comment (swist) [inline ksqldb-parser/src/test/java/io/confluent/ksql/util/QueryAnonymizerTest.java:46]: What happened here?

File: ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryAnonymizer.java [hunk 2]
```diff
@@ -461,8 +476,13 @@ public String visitSingleExpression(final SingleExpressionContext context) {
     public String visitBooleanDefault(final BooleanDefaultContext context) {
       final String columnName = context.getChild(0).getChild(0).getText();
       final String anonColumnName = getAnonColumnName(columnName);
-      final String anonValue = visit(context.getChild(0).getChild(1));
-      return String.format("%1$s=%2$s", anonColumnName, anonValue);
+
+      if (context.getChild(0).getChild(1) != null) {
```
Comment (swist) [inline ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryAnonymizer.java:480]: Don't you need to iterate over all or recurse?

File: ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/QueryAnonymizerTest.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+
+  @Test
+  public void anonQuerySplitByWordsHasOnlyTokensInSetIntersectionWithQuery() throws Exception {
+    // Given:
+    final Set<String> statements = new HashSet<>();
+    testCases
+        .filter(statement -> !statement.expectedException().isPresent())
+        .map(TestCase::statements)
+        .forEach(statements::addAll);
```
Comment (swist) [inline ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/QueryAnonymizerTest.java:65]:
```suggestion
    final Set<String> statements = testCases
        .filter(statement -> !statement.expectedException().isPresent())
        .map(TestCase::statements)
        .flatMap(Collection::stream)
        .collect(Collectors.toSet());
```

File: ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/QueryAnonymizerTest.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      intersection.retainAll(anonStatementWord);
+
+      // Assert:
+      Set<String> test = new HashSet<>(intersection);
+      intersection.removeAll(sqlTokens);
+      if (intersection.size() > 0) {
+        throw new Exception(String.format("Expected: %s\n Original: %s\n Wrong Tokens: %s",
+            statement, anonStatement, intersection.toString()));
+      }
+    }
```
Comment (swist) [inline ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/QueryAnonymizerTest.java:82]: This test is going to be hard to debug, you should look at parametrizing it:
https://www.guru99.com/junit-parameterized-test.html
(First google result so I don't know how good)

File: ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/QueryAnonymizerTest.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      final Set<String> intersection = new HashSet<>(statementWord);
+      intersection.retainAll(anonStatementWord);
+
+      // Assert:
+      Set<String> test = new HashSet<>(intersection);
+      intersection.removeAll(sqlTokens);
+      if (intersection.size() > 0) {
+        throw new Exception(String.format("Expected: %s\n Original: %s\n Wrong Tokens: %s",
+            statement, anonStatement, intersection.toString()));
+      }
```
Comment (swist) [inline ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/QueryAnonymizerTest.java:81]:
```suggestion
      assertThat(intersection, is(empty()))
```

File: ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/QueryAnonymizerTest.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+      Paths.get("src/test/java/io/confluent/ksql/test/QueriesToAnonymizeTest.txt");
+  private static final Path QUERY_VALIDATION_TEST_DIR = Paths.get("query-validation-tests");
+  private final QueryAnonymizer anon = new QueryAnonymizer();
+  private final Stream<TestCase> testCases = testFileLoader().load();
+  private List<String> sqlTokens;
+
+  @Before
+  public void setUp() {
+    sqlTokens = GrammarTokenExporter.getTokens();
+    sqlTokens.addAll(Arrays.asList("", "INT", "DOUBLE", "VARCHAR", "BOOLEAN", "BIGINT", "VALUE", "KEY"));
```
Comment (swist) [inline ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/QueryAnonymizerTest.java:34]: Also add lowercase version of all tokens. Why are you adding empty string as a token?

File: ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/QueryAnonymizerTest.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    // anonymize statements
+    final List<String> anonStatementSelection = new ArrayList<>();
+    for (String statement : statements) {
+      anonStatementSelection.add(anon.anonymize(statement));
+    }
+
+    // Assert:
+    Approvals.verifyAll("Integration tests", anonStatementSelection);
```
Comment (swist) [inline ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/QueryAnonymizerTest.java:54]: just anonymize whole file in one go using the regular approval test patterns? Anonymizer should work whether you give it individual statements or a file full of them separated by semicolons

File: ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/QueryAnonymizerTest.java (hunk 6 above)
Comment (tolgadur) [reply ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/QueryAnonymizerTest.java:34]: The anonymizer returns all tokens capitalized so there isn't a need for the lower case versions. I removed the empty string.

File: ksqldb-parser/src/test/java/io/confluent/ksql/util/QueryAnonymizerTest.java (hunk 1 above)
Comment (tolgadur) [reply ksqldb-parser/src/test/java/io/confluent/ksql/util/QueryAnonymizerTest.java:46]: Removed it.

**Review by swist — APPROVED** [review]
Changes

**Pull request merged**